File: src/com/timepath/hl2/io/bsp/VBSP.java
package com.timepath.hl2.io.bsp;

import com.timepath.hl2.io.bsp.lump.Edge;
import com.timepath.hl2.io.bsp.lump.Face;
import com.timepath.hl2.io.bsp.lump.LumpType;
import com.timepath.steam.io.storage.ACF;
import com.timepath.vfs.ZipFS;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.FloatBuffer;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author TimePath
 */
public class VBSP extends BSP {

    private static final Logger LOG = Logger.getLogger(VBSP.class.getName());

    public static void main(String[] args) throws Exception {
        BSP b = BSP.load(ACF.fromManifest(440).get("tf/maps/ctf_2fort.bsp").stream());
        System.out.println(b.getRevision());
        String ents = b.getLump(LumpType.LUMP_ENTITIES);
//        System.out.println(ents);
        ZipFS z = b.getLump(LumpType.LUMP_PAKFILE);
        System.out.println(z.name);
        b.getLump(LumpType.LUMP_VERTEXES);
    }

    @SuppressWarnings("empty-statement")
    public void processBasic() throws IOException {
        List<Float> vertices = new LinkedList<Float>();
        FloatBuffer bspVertices = getLump(LumpType.LUMP_VERTEXES);
        for(; bspVertices.hasRemaining(); vertices.add(bspVertices.get()));
        done(vertices, null);
    }

    /**
     * https://github.com/toji/webgl-source/blob/master/js/source-bsp.js#L567
     * <p/>
     * @throws java.io.IOException
     */
    public void processToji() throws IOException {
        FloatBuffer bspVertices = getLump(LumpType.LUMP_VERTEXES);
        LOG.log(Level.INFO, "Vertices: {0}", bspVertices.capacity());

        Edge[] bspEdges = getLump(LumpType.LUMP_EDGES);
        LOG.log(Level.INFO, "Edges: {0}", bspEdges.length);

        int[] bspSurfEdges = getLump(LumpType.LUMP_SURFEDGES);
        LOG.log(Level.INFO, "Surfedges: {0}", bspSurfEdges.length);

        Face[] bspFaces = getLump(LumpType.LUMP_FACES);
        LOG.log(Level.INFO, "Faces: {0}", bspFaces.length);

        List<Float> vertices = new LinkedList<Float>();
        List<Integer> indices = new LinkedList<Integer>();
        int vertexBase = 0;
        int rootPoint = 0, pointA, pointB;
        int rootVertId = 0, vertId;
        Face face;

        for(Face bspFace : bspFaces) {
            face = bspFace;
            int edgeId = face.firstedge;

            Map<Integer, Integer> vertLookupTable = new HashMap<Integer, Integer>(0);
            for(int i = 0; i < face.numedges; i++) {
                int surfEdge = bspSurfEdges[edgeId + 1];
                Edge edge = bspEdges[Math.abs(surfEdge)];
                boolean reverse = (surfEdge >= 0);

                if(i == 0) {
                    rootVertId = edge.v[reverse ? 0 : 1];
                    rootPoint = this.compileGpuVertex(bspVertices, rootVertId, vertices);
                    vertLookupTable.put(rootVertId, rootPoint);

                    vertId = edge.v[reverse ? 1 : 0];
                    pointB = this.compileGpuVertex(bspVertices, vertId, vertices);
                    vertLookupTable.put(vertId, pointB);
                } else {
                    vertId = edge.v[reverse ? 0 : 1];
                    if(vertId == rootVertId) {
                        continue;
                    }
                    if(vertLookupTable.containsKey(vertId)) {
                        pointA = vertLookupTable.get(vertId);
                    } else {
                        pointA = this.compileGpuVertex(bspVertices, vertId, vertices);
                        vertLookupTable.put(vertId, pointA);
                    }

                    vertId = edge.v[reverse ? 1 : 0];
                    if(vertId == rootVertId) {
                        continue;
                    }
                    if(vertLookupTable.containsKey(vertId)) {
                        pointB = vertLookupTable.get(vertId);
                    } else {
                        pointB = this.compileGpuVertex(bspVertices, vertId, vertices);
                        vertLookupTable.put(vertId, pointB);
                    }
                    indices.add(rootPoint - vertexBase);
                    indices.add(pointA - vertexBase);
                    indices.add(pointB - vertexBase);
                }
            }
        }
        done(vertices, indices);
    }

    /**
     * https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L222
     * <p/>
     * @throws java.io.IOException
     */
    @SuppressWarnings("empty-statement")
    public void processW23() throws IOException {
        FloatBuffer bspVertices = getLump(LumpType.LUMP_VERTEXES);
        LOG.log(Level.INFO, "Vertices: {0}", bspVertices.capacity());

        Edge[] bspEdges = getLump(LumpType.LUMP_EDGES);
        LOG.log(Level.INFO, "Edges: {0}", bspEdges.length);

        int[] bspSurfEdges = getLump(LumpType.LUMP_SURFEDGES);
        LOG.log(Level.INFO, "Surfedges: {0}", bspSurfEdges.length);

        // https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L253
        for(int i = 0; i < bspSurfEdges.length; i++) {
            if(bspSurfEdges[i] >= 0) {
                bspSurfEdges[i] = bspEdges[bspSurfEdges[i]].v[0];
            } else {
                bspSurfEdges[i] = bspEdges[-bspSurfEdges[i]].v[1];
            }
        }

        Face[] bspFaces = getLump(LumpType.LUMP_FACES);
        LOG.log(Level.INFO, "Faces: {0}", bspFaces.length);

        List<Float> vertices = new LinkedList<Float>();
        List<Integer> indices = new LinkedList<Integer>();

        for(Face face : bspFaces) {
            int edgeId = face.firstedge;

            // https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L347
            int index_shift = vertices.size();

            for(int i = 0; i < face.numedges; i++) {
                bspVertices.position(bspSurfEdges[edgeId + i] * 3);
                for(int j = 0; j++ < 3; vertices.add(bspVertices.get()));
                if(i >= 2) {
                    indices.add(index_shift + 0);
                    indices.add(index_shift + i - 1);
                    indices.add(index_shift + i);
                }
            }
        }
        done(vertices, indices);
    }

    @SuppressWarnings("empty-statement")
    private int compileGpuVertex(FloatBuffer verts, int pos, List<Float> vertices) {
        int index = vertices.size() / 3;
        verts.position(pos);
        for(int i = 0; i++ < 3; vertices.add(verts.get()));
        return index;
    }

    private void done(List<Float> vertices, List<Integer> indices) {
        if(indices != null && !indices.isEmpty()) {
            indexBuffer = ByteBuffer.allocateDirect(indices.size() * 4).asIntBuffer();
            for(int i : indices) {
                indexBuffer.put(i);
            }
            indexBuffer.flip();
            LOG.log(Level.INFO, "Map: indices {0}, triangles {1}", new Object[] {indices.size(), indices.size() / 3});
        }

        if(vertices != null && !vertices.isEmpty()) {
            vertexBuffer = ByteBuffer.allocateDirect(vertices.size() * 4).asFloatBuffer();
            for(float v : vertices) {
                vertexBuffer.put(v);
            }
            vertexBuffer.flip();
            LOG.log(Level.INFO, "Map: vertices {0}", new Object[] {vertices.size()});
        }
    }

    @Override
    protected void process() throws IOException {
        processBasic();
    }

}

File: src/main/kotlin/com/timepath/hl2/io/bsp/VBSP.kt
package com.timepath.hl2.io.bsp

import com.timepath.Logger
import com.timepath.hl2.io.bsp.lump.Edge
import com.timepath.hl2.io.bsp.lump.Face
import com.timepath.hl2.io.bsp.lump.LumpType
import com.timepath.io.OrderedInputStream
import com.timepath.steam.io.storage.ACF
import com.timepath.vfs.provider.zip.ZipFileProvider
import java.nio.ByteBuffer
import java.nio.FloatBuffer
import java.util.HashMap
import java.util.LinkedList
import kotlin.platform.platformStatic

public class VBSP(header: BSP.Header, input: OrderedInputStream) : BSP(header, input) {

    override fun process() = processBasic()

    private fun done(vertices: List<Float>?, indices: List<Int>?) {
        if ((indices != null) && indices.isNotEmpty()) {
            val new = ByteBuffer.allocateDirect(indices.size() * 4).asIntBuffer()
            indices.forEach { new.put(it) }
            new.flip()
            this.indices = new
            LOG.info { "Map: indices ${indices.size()}, triangles ${indices.size() / 3}" }
        }
        if ((vertices != null) && !vertices.isEmpty()) {
            val new = ByteBuffer.allocateDirect(vertices.size() * 4).asFloatBuffer()
            vertices.forEach { new.put(it) }
            new.flip()
            this.vertices = new
            LOG.info { "Map: vertices ${vertices.size()}" }
        }
    }

    fun processBasic() {
        val vertices = LinkedList<Float>()
        val bspVertices = getLump<FloatBuffer>(LumpType.LUMP_VERTEXES)
        while (bspVertices!!.hasRemaining()) {
            vertices.add(bspVertices.get())
        }
        done(vertices, null)
    }

    /** https://github.com/toji/webgl-source/blob/master/js/source-bsp.js#L567 */
    public fun processToji() {
        val bspVertices = getLump<FloatBuffer>(LumpType.LUMP_VERTEXES)!!
        LOG.info { "Vertices: ${bspVertices.capacity()}" }
        val bspEdges = getLump<Array<Edge>>(LumpType.LUMP_EDGES)!!
        LOG.info { "Edges: ${bspEdges.size()}" }
        val bspSurfEdges = getLump<IntArray>(LumpType.LUMP_SURFEDGES)!!
        LOG.info { "Surfedges: ${bspSurfEdges.size()}" }
        val bspFaces = getLump<Array<Face>>(LumpType.LUMP_FACES)!!
        LOG.info { "Faces: ${bspFaces.size()}" }
        val vertices = LinkedList<Float>()
        val indices = LinkedList<Int>()
        val vertexBase = 0
        var rootPoint = 0
        var rootVertId = 0

        fun compileGpuVertex(verts: FloatBuffer, pos: Int, vertices: MutableCollection<Float>): Int {
            val index = vertices.size() / 3
            verts.position(pos)
            repeat(3) { vertices.add(verts.get()) }
            return index
        }

        for (face in bspFaces) {
            val edgeId = face.firstedge
            val vertLookupTable = HashMap<Int, Int>(0)
            for (i in 0..face.numedges.toInt() - 1) {
                val surfEdge = bspSurfEdges[edgeId + 1]
                val edge = bspEdges[Math.abs(surfEdge)]
                val reverse = surfEdge >= 0
                var vertId: Int
                val pointB: Int
                if (i == 0) {
                    rootVertId = edge.v[if (reverse) 0 else 1].toInt()
                    rootPoint = compileGpuVertex(bspVertices, rootVertId, vertices)
                    vertLookupTable.put(rootVertId, rootPoint)
                    vertId = edge.v[if (reverse) 1 else 0].toInt()
                    pointB = compileGpuVertex(bspVertices, vertId, vertices)
                    vertLookupTable.put(vertId, pointB)
                } else {
                    vertId = edge.v[if (reverse) 0 else 1].toInt()
                    if (vertId == rootVertId) {
                        continue
                    }
                    val pointA: Int
                    if (vertLookupTable.containsKey(vertId)) {
                        pointA = vertLookupTable[vertId]
                    } else {
                        pointA = compileGpuVertex(bspVertices, vertId, vertices)
                        vertLookupTable.put(vertId, pointA)
                    }
                    vertId = edge.v[if (reverse) 1 else 0].toInt()
                    if (vertId == rootVertId) {
                        continue
                    }
                    if (vertLookupTable.containsKey(vertId)) {
                        pointB = vertLookupTable[vertId]
                    } else {
                        pointB = compileGpuVertex(bspVertices, vertId, vertices)
                        vertLookupTable.put(vertId, pointB)
                    }
                    indices.add(rootPoint - vertexBase)
                    indices.add(pointA - vertexBase)
                    indices.add(pointB - vertexBase)
                }
            }
        }
        done(vertices, indices)
    }

    /** https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L222 */
    public fun processW23() {
        val bspVertices = getLump<FloatBuffer>(LumpType.LUMP_VERTEXES)!!
        LOG.info({ "Vertices: ${bspVertices.capacity()}" })
        val bspEdges = getLump<Array<Edge>>(LumpType.LUMP_EDGES)!!
        LOG.info({ "Edges: ${bspEdges.size()}" })
        val bspSurfEdges = getLump<IntArray>(LumpType.LUMP_SURFEDGES)!!
        LOG.info({ "Surfedges: ${bspSurfEdges.size()}" })
        // https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L253
        for (i in bspSurfEdges.indices) {
            bspSurfEdges[i] = (if (bspSurfEdges[i] >= 0) bspEdges[bspSurfEdges[i]].v[0] else bspEdges[-bspSurfEdges[i]].v[1]).toInt()
        }
        val bspFaces = getLump<Array<Face>>(LumpType.LUMP_FACES)!!
        LOG.info({ "Faces: ${bspFaces.size()}" })
        val vertices = LinkedList<Float>()
        val indices = LinkedList<Int>()
        for (face in bspFaces) {
            val edgeId = face.firstedge
            // https://github.com/w23/OpenSource/blob/master/src/BSP.cpp#L347
            val index_shift = vertices.size()
            for (i in 0..face.numedges.toInt() - 1) {
                bspVertices.position(bspSurfEdges[edgeId + i] * 3)
                run {
                    var j = 0
                    while (j++ < 3) {
                        vertices.add(bspVertices.get())
                    }
                }
                if (i >= 2) {
                    indices.add(index_shift)
                    indices.add((index_shift + i) - 1)
                    indices.add(index_shift + i)
                }
            }
        }
        done(vertices, indices)
    }

    companion object {

        private val LOG = Logger()

        public platformStatic fun main(args: Array<String>) {
            val map = BSP.load(ACF.fromManifest(440).query("tf/maps/ctf_2fort.bsp")!!.openStream()!!)!!
            LOG.info { "Revision: ${map.revision}" }
            map.getLump<String>(LumpType.LUMP_ENTITIES)?.let {
                LOG.info { it }
            }
            map.getLump<ZipFileProvider>(LumpType.LUMP_PAKFILE)?.let {
                LOG.info { it.name }
            }
            map.getLump<Any>(LumpType.LUMP_VERTEXES)
        }

    }
}
